src/java/htsjdk/samtools/cram/build/CompressionHeaderFactory.java

```diff
@@ -60,745 +47,543 @@
+                    continue;
+                }
+                final byte[] valueBytes = tag.getValueAsByteArray();
+                try {
+                    baosForTagValues.write(valueBytes);
+                } catch (final IOException e) {
+                    throw new RuntimeIOException(e);
+                }
+            }
         }
 
-        { // mate bit flags
-            final HuffmanParamsCalculator calculator = new HuffmanParamsCalculator();
-            for (final CramCompressionRecord record : records)
-                calculator.add(record.getMateFlags());
-            calculator.calculate();
+        return baosForTagValues.toByteArray();
+    }
 
-            header.encodingMap.put(EncodingKey.MF_MateBitFlags,
-                    HuffmanIntegerEncoding.toParam(calculator.values, calculator.bitLens));
-        }
+    static ByteSizeRange geByteSizeRangeOfTagValues(final List<CramCompressionRecord> records, final int tagID) {
+        final byte type = getTagType(tagID);
+        final ByteSizeRange stats = new ByteSizeRange();
+        for (final CramCompressionRecord record : records) {
+            if (record.tags == null) {
+                continue;
+            }
 
-        { // next fragment ref id:
-            final HuffmanParamsCalculator calculator = new HuffmanParamsCalculator();
-            for (final CramCompressionRecord record : records)
-                if (record.isDetached())
-                    calculator.add(record.mateSequenceID);
-            calculator.calculate();
-
-            if (calculator.values.length == 0)
-                header.encodingMap.put(EncodingKey.NS_NextFragmentReferenceSequenceID, NullEncoding.toParam());
-
-            header.encodingMap.put(EncodingKey.NS_NextFragmentReferenceSequenceID,
-                    HuffmanIntegerEncoding.toParam(calculator.values(),
-                            calculator.bitLens()));
-            log.debug("NS: "
-                    + header.encodingMap.get(EncodingKey.NS_NextFragmentReferenceSequenceID));
+            for (final ReadTag tag : record.tags) {
+                if (tag.keyType3BytesAsInt != tagID) {
+                    continue;
+                }
+                final int size = getTagValueByteSize(type, tag.getValue());
+                if (stats.min > size)
+                    stats.min = size;
+                if (stats.max < size)
+                    stats.max = size;
+            }
         }
+        return stats;
+    }
 
-        { // next fragment alignment start
-            header.encodingMap.put(EncodingKey.NP_NextFragmentAlignmentStart, ExternalIntegerEncoding.toParam(mateInfoID));
+    /**
+     * Find a byte value never mentioned in the array
+     * @param array bytes
+     * @return byte value or -1 if the array contains all possible byte values.
+     */
+    static int getUnusedByte(final byte[] array) {
+        final byte[] usage = new byte[256];
+        for (final byte b : array) {
```

Comment:
The size of the usage array should really be based on the length if the input array. If 256 is really required, we should probably check to ensure the input array is that size.